src/policy/gc_work.rs

```diff
@@ -0,0 +1,40 @@
+/// Used to identify the trace if a policy has different kinds of traces. For example, defrag vs fast trace for Immix,
+/// mark vs forward trace for mark compact.
+pub(crate) type TraceKind = u8;
+
+pub const DEFAULT_TRACE: u8 = u8::MAX;
+
+use crate::plan::TransitiveClosure;
+use crate::scheduler::GCWorker;
+use crate::util::copy::CopySemantics;
+
+use crate::util::ObjectReference;
+
+use crate::vm::VMBinding;
+
+/// This trait defines policy-specific behavior for tracing objects.
+/// The procedural macro #[derive(PlanTraceObject)] will generate code
+/// that uses this trait. We expect any policy to implement this trait.
+/// For the sake of performance, the implementation
+/// of this trait should mark methods as `[inline(always)]`.
+pub trait PolicyTraceObject<VM: VMBinding> {
+    /// Trace object in the policy. If the policy copies objects, we should
+    /// expect `copy` to be a `Some` value.
+    fn trace_object<T: TransitiveClosure, const KIND: TraceKind>(
```

Comment:
> `TraceKind` is only for immix. I did not intend to make it general. It was used specifically for immix to reduce code duplication instead of creating two ProcessEdges packets.

Using `TraceKind` can also reduce code duplication for mark compact, which does mark trace and forward trace.

> Also, different policies or plans may use different variations of `trace_object`, with different parameters and behaviours. Unifying the interface and using TraceKind will add a cost for a policy to add a different `trace_object` -- it has to modify this interface and change other policies' code.

I am hoping that this `trace_object()` could work for future policies. But if it does not work, we can expand the `trace_object()` in this trait to support it. Alternatively, if it is really necessary, a policy can choose not to implement this trait -- the only issue is that a plan with that policy cannot use `PlanTraceObject` and `PlanProcessEdges`.

> I suggest either not generalizing trace_object, or adding a way for different ProcessEdges to be able to call different trace_object functions.

A plan can implements `ProcessEdgesWork` themselves.